Tribler/Core/Utilities/utilities.py

```diff
@@ -109,94 +179,164 @@ def validTorrentFile(metainfo):
             if not isinstance(l, IntType) and not isinstance(l, LongType):
                 raise ValueError('info files length is not int, but ' + repr(type(l)))
+    else:
+        raise ValueError("neither length or files found in files information")
+    return None
+
+
+def validate_torrent_info(metainfo):
+    """
+    Validate the info in the metainfo if it exists.
+
+    The following information is validated:
+        -   Whether an info key exists within metainfo
+        -   Whether necessary keys exist within information (name, piece length and either root hash or pieces)
+        -   Whether the name value is of type String
+        -   Whether the piece length value is of type Long or Integer
+        -   Whether the root hash is an instance of String qnd has a length of 20 if it exists
+        -   Whether the pieces value is an instance of String and it has a length of multiple of 20
+
+    The validate_files function is used to validate more as well.
+
+    :param metainfo: the metainfo for which the information has to be validated
+    :return: info if all validations succeed
+    :raise ValueError: if one of the above validations do not succeed
+    """
+    if 'info' not in metainfo:
+        raise ValueError('metainfo misses key info')
+
+    info = metainfo['info']
+    if not isinstance(info, DictType):
+        raise ValueError('info not dict')
+
+    if not(all(x in info for x in {'name', 'piece length'}) and any(x in info for x in {'root hash', 'pieces'})):
+        raise ValueError('info misses key')
+
+    name = info['name']
+    if not isinstance(name, StringType):
+        raise ValueError('info name is not string but ' + repr(type(name)))
+
+    pl = info['piece length']
+    if not (isinstance(pl, IntType) or isinstance(pl, LongType)):
+        raise ValueError('info piece size is not int, but ' + repr(type(pl)))
+
+    if 'root hash' in info:
+        rh = info['root hash']
+        if not isinstance(rh, StringType) or len(rh) != 20:
+            raise ValueError('info roothash is not 20-byte string')
+    else:
+        p = info['pieces']
+        if not isinstance(p, StringType) or len(p) % 20 != 0:
+            raise ValueError('info pieces is not multiple of 20 bytes')
+
+    validate_files(info)
+
+    return info
+
+
+def create_valid_metainfo(metainfo):
+    """
+    Creates a valid metainfo dictionary by validating the elements and correcting when possible.
+
+    :param metainfo: the metainfo that has to be validated
+    :return: the original metainfo with corrected elements if possible
+    :raise ValueError: if there is a faulty element which cannot be corrected.
+    """
+    metainfo_result = metainfo
+
+    if not isinstance(metainfo, DictType):
+        raise ValueError('metainfo not dict')
+
+    # some .torrent files have a dht:// url in the announce field.
+    if 'announce' in metainfo and not (is_valid_url(metainfo['announce']) or metainfo['announce'].startswith('dht:')):
```

Comment:
Please use parenthesis to separate the and/or from each other and to make the order of evaluation more clear.